packages/utils/telemetry-utils/src/layerCompatError.ts

```diff
@@ -0,0 +1,60 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	checkLayerCompatibility,
+	type FluidLayer,
+	type ILayerCompatDetails,
+	type ILayerCompatSupportRequirements,
+} from "@fluid-internal/client-utils";
+import type { IErrorBase, ITelemetryBaseLogger } from "@fluidframework/core-interfaces";
+
+import { UsageError } from "./error.js";
+
+/**
+ * Validates the compatibility between two layers using their compatibility details and support requirements.
+ * @internal
+ */
+export function validateLayerCompatibility(
+	layer1: FluidLayer,
+	layer2: FluidLayer,
+	compatDetailsLayer1: Pick<ILayerCompatDetails, "pkgVersion" | "generation">,
+	compatSupportRequirementsLayer1: ILayerCompatSupportRequirements,
+	maybeCompatDetailsLayer2: ILayerCompatDetails | undefined,
+	disposeFn: (error?: IErrorBase) => void,
+	logger: ITelemetryBaseLogger,
+): void {
+	const layerCheckResult = checkLayerCompatibility(
+		compatSupportRequirementsLayer1,
+		maybeCompatDetailsLayer2,
+	);
+	if (!layerCheckResult.isCompatible) {
+		const coreProperties = {
+			layer: layer1,
+			incompatibleLayer: layer2,
+			[`${layer1}Version`]: compatDetailsLayer1.pkgVersion,
+			[`${layer2}Version`]: maybeCompatDetailsLayer2?.pkgVersion ?? "unknown",
+			diff: compatDetailsLayer1.generation - (maybeCompatDetailsLayer2?.generation ?? 0),
```

Comment:
Have you considered making `diff` a unique value (i.e. undefined) if layer 2 doesn't have a `generation` value? Although I think it makes sense to consider any layer that predates generation logic to be "generation 0".